### PR TITLE
Enforce SetsRequiredMembers nullability on implicit chained constructors

### PR DESCRIPTION
**Pull request by 333fred**
Fixes https://github.com/dotnet/roslyn/issues/63768. I did the refactoring I probably should have done in the previous PR, and always pass through the initializer clause being used by the constructor under analysis. This means that we need to synthesize that initializer clause earlier, and during nullable analysis in the semantic model where we previously didn't need to. This allows to always know what the precise initializer being analyzed is, regardless of what syntax form is being passed along, and regardless of whether it is an implicit base call or not.

### CODE REVIEW

**Comment by 333fred** [comment]
@RikkiGibson @dotnet/roslyn-compiler for review.

File: src/Compilers/CSharp/Test/Symbol/Symbols/RequiredMembersTests.cs [hunk 1]
```diff
@@ -4431,6 +4479,139 @@ public record Derived(bool unused) : Base(unused);
         );
     }
 
+    [Fact, CompilerTrait(CompilerFeature.NullableReferenceTypes)]
+    public void RequiredMemberSuppressesNullabilityWarnings_ChainedConstructor_11()
```
Comment (333fred) [inline src/Compilers/CSharp/Test/Symbol/Symbols/RequiredMembersTests.cs:4483]: This test is unchanged: just moving it back to be in line with the rest of the tests with the same name.

**Review by RikkiGibson — APPROVED** [review]

File: src/Compilers/CSharp/Portable/Compiler/MethodCompiler.cs [hunk 2]
```diff
@@ -1780,6 +1781,11 @@ public override object VisitField(FieldSymbol symbol, TypeCompilationState argum
                     NullableWalker.SnapshotManager? snapshotManager = null;
                     ImmutableDictionary<Symbol, Symbol>? remappedSymbols = null;
                     var compilation = bodyBinder.Compilation;
+                    constructorInitializer = methodBody is BoundConstructorMethodBody { Initializer: { } initializer }
```
[the remark this answers is not in the document]
Comment (333fred) [reply src/Compilers/CSharp/Portable/Compiler/MethodCompiler.cs:1784]: > It feels like it might be simpler to bind the implicit constructor initializer when the BoundConstructorMethodBody is created. In effect, make it so every BoundConstructorMethodBody for an instance constructor has an initializer.

I did think about this, but it will change the observable results in APIs such as IOperation. While I'm probably ok with making such a change, I'd want it to have dedicated tests and I don't think this PR is the place to make that change.

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 3]
```diff
@@ -1349,7 +1347,7 @@ private void EnforceDoesNotReturn(SyntaxNode? syntaxOpt)
                 ownsDiagnostics = false;
             }
 
-            MethodSymbol? baseOrThisInitializer = GetConstructorThisOrBaseSymbol(constructorBody);
+            MethodSymbol? baseOrThisInitializer = constructorBody is BoundConstructorMethodBody { Initializer: BoundStatement initializer } ? initializer.GetConstructorInitializerThisOrBaseSymbol() : null;
```
[the remark this answers is not in the document]
Comment (333fred) [reply src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:1350]: Because I forgot to unrefactor.

**Comment by 333fred** [comment]
@dotnet/roslyn-compiler for a second review please.

File: src/Compilers/CSharp/Portable/Compiler/MethodCompiler.cs [hunk 4]
```diff
@@ -1912,7 +1926,11 @@ public override object VisitField(FieldSymbol symbol, TypeCompilationState argum
                 return MethodBodySynthesizer.ConstructDestructorBody(method, body);
             }
 
-            var constructorInitializer = BindImplicitConstructorInitializerIfAny(method, compilationState, diagnostics);
+            if (!attemptedInitializerBinding)
+            {
+                constructorInitializer ??= BindImplicitConstructorInitializerIfAny(method, compilationState, diagnostics);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Compiler/MethodCompiler.cs:1931]: >constructorInitializer [](http://example.com/codeflow?start=16&length=22)

Can this be not null here?

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 5]
```diff
@@ -1060,7 +1058,7 @@ int getSlotForFieldOrPropertyOrEvent(Symbol member)
         /// </summary>
         private MethodSymbol? GetBaseOrThisInitializer()
         {
-            return (_baseOrThisInitializer ?? GetConstructorThisOrBaseSymbol(this.methodMainNode));
+            return _baseOrThisInitializer ?? (this.methodMainNode is BoundConstructorMethodBody { Initializer: BoundStatement initializer } ? initializer.GetConstructorInitializerThisOrBaseSymbol() : null);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:1061]: > _baseOrThisInitializer ?? (this.methodMainNode is BoundConstructorMethodBody { Initializer: BoundStatement initializer } ? initializer.GetConstructorInitializerThisOrBaseSymbol() : null); [](http://example.com/codeflow?start=18&length=188)

Is this change necessary?

File: src/Compilers/CSharp/Portable/Compiler/MethodCompiler.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Compiler/MethodCompiler.cs:1785]: > While I'm probably ok with making such a change, I'd want it to have dedicated tests and I don't think this PR is the place to make that change.

In that case, I would like to suggest to implement suggested refactoring in a separate PR, get it merged, and then come back to addressing the problem we are trying to address in this PR.

**Comment by 333fred** [comment]
Superseded by https://github.com/dotnet/roslyn/pull/64408.
